src/Microsoft.DotNet.Build.VstsBuildsApi/VstsBuildClient.cs

```diff
@@ -0,0 +1,237 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. 
+// The .NET Foundation licenses this file to you under the MIT license. 
+// See the LICENSE file in the project root for more information. 
+
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace VstsBuildsApi
+{
+    public class VstsBuildClient
+    {
+        private string _collectionIdentifier;
+        private string _credentials;
+
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        /// <param name="collectionIdentifier">collection identifer for build definitions</param>
+        /// <param name="credentials">credentials used for basic rest api authentication</param>
+        public VstsBuildClient(string collectionIdentifier, string credentials)
+        {
+            if(string.IsNullOrWhiteSpace(collectionIdentifier))
+            {
+                throw new Exception("Required parameter (collectionIdentifier) cannot be null or white space.");
+            }
+            if(!Uri.IsWellFormedUriString(collectionIdentifier, UriKind.Relative))
+            {
+                throw new Exception(string.Format("collectionIdentifier '{0}' contains invalid characters.  collectionIdentifier must contain only valid uri characters.", collectionIdentifier));
+            }
+
+            _collectionIdentifier = collectionIdentifier;
+            _credentials = credentials;
+        }
+
+        /// <summary>
+        /// Load a local build definition, combine the uniqueIdentifier with the definition name, 
+        /// check VSTS for a matching definition name, if present, update that definition with the
+        /// local build definition, otherwise create a new definition.
+        /// </summary>
+        /// <param name="stream">Stream to a VSTS build definition</param>
+        /// <returns>Created or updated build Id</returns>
+        public async Task<string> CreateOrUpdateDefinitionAsync(Stream stream)
+        {
+            JObject definition = GetDefinition(stream);
+            return await CreateOrUpdateDefinitionAsync(definition).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Load a local build definition, combine the uniqueIdentifier with the definition name, 
+        /// check VSTS for a matching definition name, if present, update that definition with the
+        /// local build definition, otherwise create a new definition.
+        /// </summary>
+        /// <param name="definition">VSTS build definition JSON object</param>
+        /// <returns>Created or updated build definition id</returns>
+        private async Task<string> CreateOrUpdateDefinitionAsync(JObject definition)
+        {
+            var key = string.Join("_", _collectionIdentifier, definition["name"].ToString());
```

Comment:
Why are you using Join to concat 2 strings separated by a "_"? I would just concat them `_collectionIdentifier + "_" + definition["name"].ToString()`